### ISSUE TITLE
Tar.Header.Header.unmarshal should treat `Normal` entries with trailing slash name as `Directory`.

### ISSUE THREAD

**Issue by gravicappa** [mirage/ocaml-tar#129]
As written [here](https://www.gnu.org/software/tar/manual/html_node/Standard.html).
> Also, for backward compatibility, tar treats a regular file whose name ends with a slash as a directory.

And I have encountered archives like that.

**Comment by gravicappa** [comment]
You can create such archives with `bsdtar c --format v7 …`.

I would implement that `link_indicator` fix in `Tar.Header.unmarshal` either when `hdr_version` (not `level` though) is pre-ustar or unconditionally after synthesizing long names (to avoid being misdirected by trailing slash in cut-off name). In my code I just call
```
let sanitize_header (header: Tar.Header.t) =
  match header.link_indicator with
  | Normal ->
      if path_is_directory header.file_name then
        { header with link_indicator = Directory }
      else
        header
  | _ -> header
```
in `Tar_unix.Archive.with_next_file` called proc and it generally works although I'm dealing with docker image layers archives only at the moment.

**Comment by reynir** [comment]
I was curious how gnu tar would treat a tar archive with a pax header setting the file name to `clearly/a/directory/` followed by a header for an empty file `placeholder` with normal link indicator. So I created that with ocaml-tar, and interestingly tar treats the contents as a directory `clearly/a/directory/`:
```shell
$ tar -tvf pax-shenanigans.tar 
dr-------- 0/0               0 1970-01-01 01:00 clearly/a/directory/
```
Here's the sample archive (gzipped so github accepts it) [pax-shenanigans.tar.gz](https://github.com/mirage/ocaml-tar/files/11770475/pax-shenanigans.tar.gz)

I will make a PR shortly implementing your second proposal.

**Comment by reynir** [comment]
tar.2.5.1 has been submitted to opam-repository with a fix. Thank you for the report and the discussion on how to fix it.

**Comment by gravicappa** [comment]
It seems that I was wrong: the fix up for `link_indicator` should be done in [`read_header`](https://github.com/mirage/ocaml-tar/blob/a722218b375af5a91b424b59ca7c556470242a2e/lib/tar.ml#L786) instead: `file_name` is being updated there.

And I encountered archive (which unfortunately I can't provide) there `LongLink` contains full name and `file_name` is cut off right after `/`.

**Comment by gravicappa** [comment]
This is how I fixed the issue locally (probably missed a couple of other issues):
```
diff --git a/lib/tar.ml b/lib/tar.ml
index 0ff1346..16de231 100644
--- a/lib/tar.ml
+++ b/lib/tar.ml
@@ -522,17 +522,7 @@ module Header = struct
         let mod_time = match extended.Extended.mod_time with
           | None -> get_hdr_mod_time c
           | Some x -> x in
-        let link_indicator =
-          let link_indicator = get_hdr_link_indicator c in
-          (* For backward compatibility we treat normal files ending in slash
-             as directories. Because [Link.of_char] treats unrecognized link
-             indicator values as normal files we check directly *)
-          if String.length file_name > 0 && file_name.[String.length file_name - 1] = '/' &&
-             (link_indicator = '0' || link_indicator = '\000') then
-            Link.Directory
-          else
-            Link.of_char ~level link_indicator
-        in
+        let link_indicator = Link.of_char ~level (get_hdr_link_indicator c) in
         let uname = match extended.Extended.uname with
           | None -> if ustar then get_hdr_uname c else ""
           | Some x -> x in
@@ -718,6 +708,17 @@ module HeaderReader(Async: ASYNC)(Reader: READER with type 'a t = 'a Async.t) =
           | None -> return (Error `Eof)
         end in
 
+    let true_link_indicator link_indicator file_name =
+      (* For backward compatibility we treat normal files ending in slash
+         as directories. Because [Link.of_char] treats unrecognized link
+         indicator values as normal files we check directly *)
+      if String.length file_name > 0
+         && file_name.[String.length file_name - 1] = '/'
+         && link_indicator = Header.Link.Normal then
+        Header.Link.Directory
+      else
+        link_indicator in
+
     let rec read_header (file_name, link_name, hdr) : (Header.t, [`Eof]) result Async.t =
       let raw_link_indicator = Header.get_hdr_link_indicator buffer in
       if (raw_link_indicator = 'K' || raw_link_indicator = 'L') && level = Header.GNU then
@@ -738,7 +739,8 @@ module HeaderReader(Async: ASYNC)(Reader: READER with type 'a t = 'a Async.t) =
       else begin
         let link_name = if link_name = "" then hdr.Header.link_name else link_name in
         let file_name = if file_name = "" then hdr.Header.file_name else file_name in
-        return (Ok {hdr with Header.link_name; file_name })
+        let link_indicator = true_link_indicator hdr.Header.link_indicator file_name in
+        return (Ok {hdr with Header.link_name; file_name; link_indicator })
       end in
     get_hdr ()
     >>= function
```

**Comment by hannesm** [comment]
reopening, thanks for your testing and reporting!

**Comment by reynir** [comment]
Dear @gravicappa, thank you for your report and for sharing your own fix, and sorry for not reacting earlier.

Unfortunately `Link.of_char` returns `Link.Normal` if the link indicator is unknown - so maybe a bit too much is treated as directories. Ideally, only `'0'` and `'\000'` would receive this treatment. I don't know how often these unknown-to-us link indicators occur in the wild, and I imagine the chances are slim that they end in a slash so your solution is probably acceptable. In #127 I will reconsider returning `Link.Normal` on "unknown" link indicators.

**Comment by reynir** [comment]
A fix's released in 2.6.0. I will keep this issue open until I have a fix for the `next` branch as well. Thanks again for your reports and suggestions!

**Comment by hannesm** [comment]
Done in #127
